lib/hutch/waiter.rb

```diff
@@ -14,28 +25,70 @@ def wait_until_signaled
       self.sig_read, self.sig_write = IO.pipe
 
       register_signal_handlers
-      wait_for_signal
 
-      sig = sig_read.gets.strip.downcase
-      logger.info "caught sig#{sig}, stopping hutch..."
+      begin
+        wait_for_signal
+
+        sig = sig_read.gets.strip
+        handle_signal(sig)
+      rescue ContinueProcessingSignals
+        retry
+      end
+    end
+
+    def handle_signal(sig)
+      raise ContinueProcessingSignals unless REGISTERED_SIGNALS.include?(sig)
+      if user_signal?(sig)
+        handle_user_signal(sig)
+      else
+        handle_shutdown_signal(sig)
+      end
+    end
+
+    def handle_shutdown_signal(sig)
+      logger.info "caught SIG#{sig}, stopping hutch..."
+    end
+
+    # @raises ContinueProcessingSignals
+    def handle_user_signal(sig)
+      case sig
+      when 'USR2' then handle_usr2
+      else raise "Assertion failed - unhandled signal: #{sig.inspect}"
+      end
+      raise ContinueProcessingSignals
     end
 
     private
 
+    def handle_usr2
+      Thread.list.each do |thread|
```

Comment:
I suggest that we log a header such as "requested a VM-wide thread stack trace dump…"